batch_normalization/__init__.py

```diff
@@ -0,0 +1,122 @@
+from blocks.bricks.cost import CategoricalCrossEntropy, MisclassificationRate
+from blocks.extensions import Printing
+from blocks.extensions.monitoring import DataStreamMonitoring
+from blocks.extensions import FinishAfter
+from blocks.graph import (ComputationGraph, apply_batch_normalization,
+                          get_batch_normalization_updates)
+from blocks.initialization import Constant, IsotropicGaussian
+from blocks.main_loop import MainLoop
+from fuel.datasets.toy import Spiral
+from fuel.streams import DataStream
+from fuel.schemes import ShuffledScheme, SequentialScheme
+
+
+def main(num_epochs=50, batch_normalized=True, alpha=0.1):
+    """Run the example.
+
+    Parameters
+    ----------
+    num_epochs : int, optional
+        Number of epochs for which to train.
+
+    batch_normalized : bool, optional
+        Batch-normalize the training graph. Defaults to `True`.
+
+    alpha : float, optional
+        Weight to apply to a new sample when calculating running
+        averages for population statistics (1 - alpha weight is
+        given to the existing average).
+
+    """
+    if batch_normalized:
+        # Add an extra keyword argument that only BatchNormalizedMLP takes,
+        # in order to speed things up at the cost of a bit of extra memory.
+        mlp_class = BatchNormalizedMLP
+        extra_kwargs = {'conserve_memory': False}
+    else:
+        mlp_class = MLP
+        extra_kwargs = {}
+    mlp = mlp_class([Logistic(), Logistic(), Logistic(), Softmax()],
+                    [2, 5, 5, 5, 3],
+                    weights_init=IsotropicGaussian(0.2),
+                    biases_init=Constant(0.), **extra_kwargs)
+    mlp.initialize()
+
+    # Generate a dataset with 3 spiral arms, using 8000 examples for
+    # training and 2000 for testing.
+    dataset = Spiral(num_examples=10000, classes=3,
+                     sources=['features', 'label'],
+                     noise=0.05)
+    train_stream = DataStream(dataset,
+                              iteration_scheme=ShuffledScheme(examples=8000,
+                                                              batch_size=20))
+    test_stream = DataStream(dataset,
+                             iteration_scheme=SequentialScheme(
+                                 examples=list(range(8000, 10000)),
+                                 batch_size=2000))
+
+    # Build a cost graph; this contains BatchNormalization bricks that will
+    # by default run in inference mode.
+    features = tensor.matrix('features')
+    label = tensor.lvector('label')
+    prediction = mlp.apply(features)
+    cost = CategoricalCrossEntropy().apply(label, prediction)
+    misclass = MisclassificationRate().apply(label, prediction)
+    misclass.name = 'misclass'  # The default name for this is annoyingly long
+    original_cg = ComputationGraph([cost, misclass])
+
+    if batch_normalized:
+        cg = apply_batch_normalization(original_cg)
```

Comment:
I'd explain what `apply_batch_normalization` does, i.e. take an inference graph and return a graph where population statistics have been replaced by batch statistics.